enterprise/coderd/prebuilds/claim_test.go

```diff
@@ -0,0 +1,427 @@
+package prebuilds_test
+
+import (
+	"context"
+	"database/sql"
+	"strings"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/coder/serpent"
+
+	"github.com/coder/coder/v2/coderd/coderdtest"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/dbtestutil"
+	"github.com/coder/coder/v2/coderd/rbac"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/enterprise/coderd/coderdenttest"
+	"github.com/coder/coder/v2/enterprise/coderd/license"
+	"github.com/coder/coder/v2/enterprise/coderd/prebuilds"
+	"github.com/coder/coder/v2/provisioner/echo"
+	"github.com/coder/coder/v2/provisionersdk/proto"
+	"github.com/coder/coder/v2/testutil"
+)
+
+type storeSpy struct {
+	database.Store
+
+	claims           *atomic.Int32
+	claimParams      *atomic.Pointer[database.ClaimPrebuiltWorkspaceParams]
+	claimedWorkspace *atomic.Pointer[database.ClaimPrebuiltWorkspaceRow]
+}
+
+func newStoreSpy(db database.Store) *storeSpy {
+	return &storeSpy{
+		Store:            db,
+		claims:           &atomic.Int32{},
+		claimParams:      &atomic.Pointer[database.ClaimPrebuiltWorkspaceParams]{},
+		claimedWorkspace: &atomic.Pointer[database.ClaimPrebuiltWorkspaceRow]{},
+	}
+}
+
+func (m *storeSpy) InTx(fn func(store database.Store) error, opts *database.TxOptions) error {
+	// Pass spy down into transaction store.
+	return m.Store.InTx(func(store database.Store) error {
+		spy := newStoreSpy(store)
+		spy.claims = m.claims
+		spy.claimParams = m.claimParams
+		spy.claimedWorkspace = m.claimedWorkspace
+
+		return fn(spy)
+	}, opts)
+}
+
+func (m *storeSpy) ClaimPrebuiltWorkspace(ctx context.Context, arg database.ClaimPrebuiltWorkspaceParams) (database.ClaimPrebuiltWorkspaceRow, error) {
+	m.claims.Add(1)
+	m.claimParams.Store(&arg)
+	result, err := m.Store.ClaimPrebuiltWorkspace(ctx, arg)
+	if err == nil {
+		m.claimedWorkspace.Store(&result)
+	}
+	return result, err
+}
+
+func TestClaimPrebuild(t *testing.T) {
```

Comment:
We need one or more error cases here; for example if the claim fails for whatever reason (db unavailable, etc).

Reply:
Added here: https://github.com/coder/coder/pull/17458/commits/fcdbba8ecc1892af0d94951444023b4dfcd491ac
@dannykopping wdyt is it enough?

Reply:
There's a lot of duplication both between `TestClaimPrebuild`/`TestClaimPrebuild_CheckDifferentErrors` and the also between the two check funcs within `TestClaimPrebuild_CheckDifferentErrors`.

I'd prefer it if you could incorporate the err tests into `TestClaimPrebuild` to keep things concise, but it doesn't need to block this PR right now.

Aside from that, your tests represent the _behaviour_ not the _implementation_, so a name like `ErrNoClaimablePrebuiltWorkspaces is returned` is too tightly coupled with the code. Rather name it like "no claimable prebuilt workspaces results in error" or something like that.

Reply:
okay, I'll try to do it in a follow-up PR. But it will be a lot of control flags in the test. It's usually hard to follow the logic if you have a lot of control flags which significantly change the logic.

For ex.: if flag is true - successful scenario, flag is false is unsuccessful scenario.

But let's see how it will look like in a follow-up PR.

Reply:
addressed here: https://github.com/coder/coder/pull/17567